src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php

```diff
@@ -39,23 +50,59 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
         $this->joinRelations($queryBuilder, $resourceClass);
     }
 
+    public function getMetadataProperties(string $resourceClass): array
+    {
+        $properties = [];
+
+        foreach ($this->propertyNameCollectionFactory->create($resourceClass) as $property) {
+            $properties[$property] = $this->propertyMetadataFactory->create($resourceClass, $property);
+        }
+
+        return $properties;
+    }
+
     /**
      * Left joins relations to eager load.
      *
      * @param QueryBuilder $queryBuilder
      * @param string       $resourceClass
      */
-    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass)
+    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, string $originAlias = 'o', string &$relationAlias = 'a')
     {
-        $classMetaData = $queryBuilder->getEntityManager()->getClassMetadata($resourceClass);
+        $classMetadata = $queryBuilder->getEntityManager()->getClassMetadata($resourceClass);
+        $j = 0;
+
+        foreach ($classMetadata->getAssociationNames() as $i => $association) {
+            $mapping = $classMetadata->associationMappings[$association];
+            $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $association);
+
+            if (ClassMetadataInfo::FETCH_EAGER !== $mapping['fetch'] || false === $propertyMetadata->isReadableLink()) {
+                continue;
+            }
 
-        foreach ($classMetaData->getAssociationNames() as $i => $association) {
-            $mapping = $classMetaData->associationMappings[$association];
+            $method = false === $mapping['joinColumns'][0]['nullable'] ? 'innerJoin' : 'leftJoin';
 
-            if (ClassMetadataInfo::FETCH_EAGER === $mapping['fetch']) {
-                $queryBuilder->leftJoin('o.'.$association, 'a'.$i);
-                $queryBuilder->addSelect('a'.$i);
+            $associationAlias = $relationAlias.$i;
+            $queryBuilder->{$method}($originAlias.'.'.$association, $associationAlias);
+            $select = [];
+            $targetClassMetadata = $queryBuilder->getEntityManager()->getClassMetadata($mapping['targetEntity']);
+
+            foreach ($this->getMetadataProperties($mapping['targetEntity']) as $property => $propertyMetadata) {
+                if (true === $propertyMetadata->isIdentifier()) {
+                    $select[] = $property;
+                    continue;
+                }
+
+                //the field test allows to add methods to a Resource which do not reflect real database fields
+                if (true === $targetClassMetadata->hasField($property) && true === $propertyMetadata->isReadable()) {
+                    $select[] = $property;
+                }
             }
+
+            $queryBuilder->addSelect(sprintf('partial %s.{%s}', $associationAlias, implode(',', $select)));
+
+            $relationAlias = $relationAlias.++$j;
```

Comment:
We should avoid random numbers in DQL as much as possible to trigger the Doctrine cache as often as possible.